futures-util/src/stream/once.rs

```diff
@@ -18,26 +18,28 @@ pub struct Once<T, E>(Option<Result<T, E>>);
 /// use futures_executor::block_on;
 ///
 /// # fn main() {
-/// let mut stream = stream::once::<(), _>(Err(17));
-/// assert_eq!(Err(17), block_on(stream.collect()));
+/// let mut stream = stream::once::<Result<(), _>>(Err(17));
+/// let collected: Result<Vec<_>, _> = block_on(stream.collect());
+/// assert_eq!(collected, Err(17));
 ///
-/// let mut stream = stream::once::<_, ()>(Ok(92));
-/// assert_eq!(Ok(vec![92]), block_on(stream.collect()));
+/// let mut stream = stream::once::<Result<_, ()>>(Ok(92));
+/// let collected: Result<Vec<_>, _> = block_on(stream.collect());
+/// assert_eq!(collected, Ok(vec![92]));
 /// # }
 /// ```
-pub fn once<T, E>(item: Result<T, E>) -> Once<T, E> {
-    Once(Some(item))
+pub fn once<F: IntoFuture>(item: F) -> Once<F::Future> {
+    Once(Some(item.into_future()))
 }
 
-impl<T, E> Stream for Once<T, E> {
-    type Item = T;
-    type Error = E;
+impl<F: Future> Stream for Once<F> {
+    type Item = F::Item;
+    type Error = F::Error;
 
-    fn poll_next(&mut self, _: &mut task::Context) -> Poll<Option<T>, E> {
-        match self.0.take() {
-            Some(Ok(e)) => Ok(Async::Ready(Some(e))),
-            Some(Err(e)) => Err(e),
-            None => Ok(Async::Ready(None)),
+    fn poll_next(&mut self, cx: &mut task::Context) -> Poll<Option<F::Item>, F::Error> {
+        if let Some(mut f) = self.0.take() {
+            Ok(f.poll(cx)?.map(Some))
```

Comment:
If the future was `Async::Pending`, this will remove it and yield `Ok(Async::Pending)`, and all future calls will result in `Ok(Async::Ready(None))`. In order to fix this, you have to either (a) not `take` unless the value from `poll` is `Ready`, or  (b) put the future back into `self.0`.